sel/src/Sel/SecretKey/Stream.hs

```diff
@@ -0,0 +1,571 @@
+  clearTextForeignPtr <- liftIO $ Foreign.mallocForeignPtrBytes (fromIntegral messageLength)
+  let cipherTextLen = messageLength + fromIntegral cryptoSecretStreamXChaCha20Poly1305ABytes
+  liftIO $ Foreign.withForeignPtr cipherTextForeignPtr $ \cipherTextBuffer -> do
+    liftIO $ Foreign.withForeignPtr clearTextForeignPtr $ \clearTextBuffer -> do
+      tagBuffer <- sodiumMalloc 1
+      result <-
+        cryptoSecretStreamXChaCha20Poly1305Pull
+          statePtr
+          clearTextBuffer
+          Foreign.nullPtr
+          tagBuffer
+          cipherTextBuffer
+          cipherTextLen
+          Foreign.nullPtr
+          0
+      when (result /= 0) $ throw StreamDecryptionException
+      bsPtr <- Foreign.mallocBytes (fromIntegral messageLength)
+      Foreign.copyBytes bsPtr (Foreign.castPtr clearTextBuffer) (fromIntegral messageLength)
+      BSU.unsafePackMallocCStringLen (bsPtr, fromIntegral messageLength)
+
+-- | Perform streaming decryption of a finite Linked List.
+--
+-- This function can throw 'StreamDecryptionException' if the chunk is invalid, incomplete, or corrupted.
+--
+-- @since 0.0.1.0
+decryptList :: forall m. MonadIO m => SecretKey -> Header -> [CipherText] -> m (Maybe [StrictByteString])
+decryptList secretKey header encryptedMessages =
+  decryptStream secretKey header $ \multipart -> do
+    forM encryptedMessages $ \cipherText -> do
+      decryptChunk multipart cipherText
+
+-- | A secret key of size 'cryptoSecretStreamXChaCha20Poly1305KeyBytes'.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+      -- > display secretKey == "[REDACTED]"
+    )
+    via (OpaqueInstance "[REDACTED]" SecretKey)
+
+-- | @since 0.0.1.0
+instance Eq SecretKey where
+  (SecretKey hk1) == (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
+
+-- | @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
+
+-- | > show secretKey == "[REDACTED]"
+--
+-- @since 0.0.1.0
+instance Show SecretKey where
+  show _ = "[REDACTED]"
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newSecretKey :: IO SecretKey
+newSecretKey = newSecretKeyWith cryptoSecretStreamXChaCha20Poly1305KeyGen
+
+-- | Create a 'SecretKey' from a binary 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- The input secret key, once decoded from base16, must be of length
+-- 'cryptoSecretStreamXChaCha20Poly1305KeyBytes'.
+--
+-- @since 0.0.1.0
+secretKeyFromHexByteString :: Base16 StrictByteString -> Either Text SecretKey
+secretKeyFromHexByteString hexSecretKey = unsafeDupablePerformIO $
+  case Base16.decodeBase16Untyped (Base16.extractBase16 hexSecretKey) of
```

Comment:
Sorry for digging this up from the past, but I'm a bit confused by this and just wanted to see if there's some context I'm missing.

Why do we call `decodeBase16Untyped . extractBase16` on a `Base16 StrictByteString` when we could use `decodeBase16`?

If the value is a `Base16 a`, shouldn't we assume that it's already Base16-encoded and we can use `decodeBase16`?

Reply:
You're right, this is weird. I'll go over it later. Thanks!